text/0000-posix-error-numbers.md

```diff
@@ -0,0 +1,248 @@
+avoid sign-extending instructions when returning error numbers (e.g. in a tight
+loop, or up a long call stack).
+
+```rust
+#[cfg(target_arch = "x86")]
+type NonZeroFastInt = core::num::NonZeroI32;
+
+#[cfg(target_arch = "x86_64")]
+type NonZeroFastInt = core::num::NonZeroI64;
+
+pub struct ErrorNumber(NonZeroFastInt);
+```
+
+Most of the user-facing API is in trait implementations, specifically
+`PartialEq` for the integers.
+
+```rust
+impl PartialEq<i32> for ErrorNumber { ... }
+impl PartialEq<ErrorNumber> for i32 { ... }
+
+impl PartialEq<u32> for ErrorNumber { ... }
+impl PartialEq<ErrorNumber> for u32 { ... }
+
+/* ... and so on for all {u,i}{16,32,64,size} ... */
+```
+
+The implementations of `PartialEq` may also need to be arch-specific, to avoid
+unnecessary bounds checks or conversions depending on the inner value.
+
+The POSIX standard's list of error numbers is non-exhaustive, and platforms may
+define their own error numbers. The `ErrorNumber` struct supports this through
+construction from `core::num::NonZeroI16`, which is the smallest native integer
+type I would expect someone trying to run a UNIX system on.
+
+```rust
+impl ErrorNumber {
+	pub const fn new(n: core::num::NonZeroI16) -> ErrorNumber
+}
+```
+
+Some applications need to store or transmit the value of an `ErrorNumber`. This
+proposal bounds the range of `ErrorNumber` to be no greater than that of `i64`,
+so that a non-failable conversion is possible, and also offers failable
+conversions to smaller integral types.
+
+```rust
+impl From<ErrorNumber> for i64  { ... }
+impl From<ErrorNumber> for core::num::NonZeroI64  { ... }
+
+impl TryFrom<ErrorNumber> for i32 { ... }
+impl TryFrom<ErrorNumber> for core::num::NonZeroI32 { ... }
+
+/* ... and so on for all {u,i}{16,32,64,size} ... */
+```
+
+Values of POSIX error numbers for the current platform are available as associated constants, which may be used with ergonomics similar to the C
+macros defined in `<errno.h>`. We'll want to generate these via an auxiliary
+script that can compile and run a C program to detect platform values.
+
+```rust
+macro_rules! error_number {
+	($name:ident, $matches:tt) => {
+		pub const $name: ErrorNumber = ErrorNumber::new(unsafe {
+			NonZeroI16::new_unchecked(match () $matches)
+		});
+	}
+}
+
+impl ErrorNumber {
+	error_number!(EAGAIN, {
+		#[cfg(target_os = "linux", target_arch = "x86_64")] _ => 11,
+		#[cfg(target_os = "macos")] _ => 35,
+		// ...
+		#[cfg(doc)] _ => loop {},
+	});
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal increases the API footprint of the standard library.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The dominant alternative design is to use error number constants in `libc`,
+which are of type `c_int` and therefore less ergonomic to compare. Depending on
+`libc` is also undesirable for binaries that run in constrained environments.
```

Comment:
I do not understand why this is the case. The libc crate is the equivalent of a C header file.

Reply:
I think he mean use cases like writing kernel driver which uses kernel headers.
I don't know if error constant defined by kernel header differ with libc's one.